test/dynamo/test_subclasses.py

```diff
@@ -1847,17 +1849,662 @@ def inner_compile(
+        self.assertExpectedInline(
+            normalize_gm(fw[0].print_readable(print_output=False)),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, primals_1: "f32[24]", primals_2: "f32[24]"):
+        clone: "f32[24]" = torch.ops.aten.clone.default(primals_1);  primals_1 = None
+        clone_1: "f32[24]" = torch.ops.aten.clone.default(primals_2);  primals_2 = None
+
+        view: "f32[3, 2, 4]" = torch.ops.aten.view.default(clone, [3, 2, 4]);  clone = None
+        view_1: "f32[3, 2, 4]" = torch.ops.aten.view.default(clone_1, [3, 2, 4]);  clone_1 = None
+        return (view, view_1)
+""",  # noqa: B950
+        )
+
+        self.assertExpectedInline(
+            normalize_gm(bw[0].print_readable(print_output=False)),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, tangents_1: "f32[3, 2, 4]", tangents_2: "f32[3, 2, 4]"):
+        view_2: "f32[24]" = torch.ops.aten.view.default(tangents_1, [24]);  tangents_1 = None
+        view_3: "f32[24]" = torch.ops.aten.view.default(tangents_2, [24]);  tangents_2 = None
+        return (view_2, view_3)
+""",  # noqa: B950
+        )
+
+    def test_tensor_subclass_TwoTensor_return_shape(self):
+        @torch.compile(backend="aot_eager", dynamic=True)
+        def fn(x):
+            return x.clone().view(x.shape[0] * x.shape[1])
+
+        a = torch.ones(2, 3)
+        b = a.clone()
+        tt = TwoTensor(a, b)
+        out = fn(tt)
+        self.assertEqual(tt.view(2 * 3), out)
+        self.assertEqual(out.shape, (6,))
+
+    def test_tensor_subclass_TwoTensor_nested(self):
+        @torch.compile(backend="aot_eager", dynamic=True)
+        def f(x, i, y):
+            out1 = x.sin() + i.sin() + y.sin()
+            val1 = x.shape[0] * i.shape[1] * y.shape[0]
+            return out1 * val1
+
+        i = torch.randn(2, 2, requires_grad=True)
+        x = TwoTensor(i, i.clone())
+        y = TwoTensor(x.clone(), x.clone())
+
+        out = f(x, i, y)
+
+        x_test = x.clone().detach().requires_grad_(True)
+        i_test = i.clone().detach().requires_grad_(True)
+        y_test = y.clone().detach().requires_grad_(True)
+
+        out_test = f(x_test, i_test, y_test)
+        torch.allclose(out, out_test)
+
+        out.sum().backward()
+        out_test.sum().backward()
+        torch.allclose(x.grad, x_test.grad)
+        torch.allclose(i.grad, i_test.grad)
+        torch.allclose(y.grad, y_test.grad)
+
+    def test_subclass_TwoTensor_TwoTensor_TwoTensor(self):
+        @torch.compile(backend="aot_eager", dynamic=True)
+        def f(x):
+            return x.sin()
+
+        data = torch.randn(2, 3)
+        s = TwoTensor(data, data.clone())
+        y = TwoTensor(s, s.clone())
+        z = TwoTensor(s, y)
+        out = f(z)
+        self.assertEqual(out, z.sin())
+
+    @unittest.expectedFailure
+    def test_njt_subclass_simple(self):
+        def f(nt):
+            y = nt.clone()
+            return y * y.size(0) * y.size(1)
```

Comment:
same with this one (I would expect this one to pass as long as you don't include compute on `y.size(0)` or `y.size(1)` in the function, which Joel is banning anyway here: https://github.com/pytorch/pytorch/pull/138602)